### ISSUE TITLE
EOF error when running on macOS

### ISSUE THREAD

**Issue by sebsto** [recitale/recitale#26]
I tried to install recitale v2.0.0a5 on macOS. I received an `EOFError` when I try to launch the main recitale executable, whatever option is passed to it.

Here is the stack trace 

```
(web_album) ➜  web_album recitale --version 
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/bin/recitale", line 5, in <module>
    from recitale.recitale import main
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/lib/python3.8/site-packages/recitale/recitale.py", line 25, in <module>
    from .cache import CACHE
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/lib/python3.8/site-packages/recitale/cache.py", line 83, in <module>
    CACHE = Cache(json=json)
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/lib/python3.8/site-packages/recitale/cache.py", line 37, in __init__
    self.cache = Manager().dict(cache)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 57, in Manager
    m.start()
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/managers.py", line 579, in start
    self._process.start()
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
Traceback (most recent call last):
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/bin/recitale", line 5, in <module>
    from recitale.recitale import main
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/lib/python3.8/site-packages/recitale/recitale.py", line 25, in <module>
    from .cache import CACHE
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/lib/python3.8/site-packages/recitale/cache.py", line 83, in <module>
    CACHE = Cache(json=json)
  File "/Users/stormacq/.local/share/virtualenvs/web_album-5qNmljMe/lib/python3.8/site-packages/recitale/cache.py", line 37, in __init__
    self.cache = Manager().dict(cache)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/context.py", line 57, in Manager
    m.start()
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/managers.py", line 583, in start
    self._address = reader.recv()
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/connection.py", line 250, in recv
    buf = self._recv_bytes()
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/connection.py", line 383, in _recv
    raise EOFError
EOFError
```

I tried python 3.8, python 3.9 and python 3.11
I am one Apple Silicon (M1)

**Comment by QSchulz** [comment]
Hi, thanks for the report. I do not plan to support anything outside of Linux.

However, I think this error may be fixable by doing something that would benefit the project as a whole so I'll look into it.

So, `recitale` leverages (by default) all your CPU cores (hyperthreading included) by spawning processes thanks to the `multiprocessing` module.  Unfortunately, Linux, MacOS and Windows do not use the same spawning method, c.f. https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods

As far as I could understand, the `spawn` start method will re-import modules when a new process starts. The `cache` modules has some global variable set in the module namespace which means it'll be run on import.
This variable (`CACHE`) is an instance of the `Cache` class which contains a `dict` managed by `Manager` (from multiprocessing). `Manager` (c.f. https://docs.python.org/3/library/multiprocessing.html#managers) allows to share data between processes, and is handled by a server process. Using `Manager` is unfortunately required because the global cache needs to be synchronized between all processes since all are both readers and writers to this cache. At least, this is how I implemented it, maybe there's a better/simpler way to do it but didn't find any when I looked into it.

So.. the issue is that this `Manager` server process will automatically be created upon import of the module, which isn't smart at all since it isn't required unless we actually **build**  something so we should postpone the creation of the global `CACHE` variable. This is what I'm planning to do to hopefully fix this particular issue. I'll try to replicate the setup even when running on Linux thanks to the `set_start_method`, c.f. https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method.

But considering how different the start methods for processes are between Linux and Windows/MacOS and how little I understand them, I don't plan on hunting bugs that may arise from using the default one on MacOS or Windows. We could force MacOS to use the `fork` method to align with Linux but this is deemed unsafe so I'd rather not try it at all.

**Comment by sebsto** [comment]
Thank you Quentin for the context and the explanations about the design decisions.  I am not expert in Python multithreading, I know it has been picky and historically challenging, even if multiple solutions are available now.

Coming from other programming language background, it seems crazy to spawn multiple processes to achieve higher throughput. In particular, because it brings extra complexity to access common data structures, such as the cache.

I guess you already tried to use a multi threading approach vs a multi processes one, by using the thread pools and the [`Executor`](https://docs.python.org/3/library/concurrent.futures.html) interface.  Are performances significantly different ?

I'll continue to poke around and search for a solution on macOS and will let you know here the result of my investigation.

In the meantime, I can use the docker image as soon as the fix for  https://github.com/recitale/recitale/issues/25 will be released :-)

**Comment by QSchulz** [comment]
Multithreading is only, in Python, for IO-bound tasks because multiple threads from the same Python process are all using the GIL (Global Interpreter Lock), this means that essentially it makes only sense when threads can be idle. That is not the case for `recitale` which is heavily CPU-bound with doing image thumbnailing, it's barely waiting on IO at any point in time so you'll load only one CPU-core with one-thread and that's it.

I haven't looked into `Executor` specifically, but considering there's a `ProcessPoolExecutor` specifically for side-stepping the GIL with an example using CPU intensive tasks (finding if a number is prime).

This was not that much of an issue in `prosopopee` because it was spawning a subprocess calling `gm` which is itself running the tasks in parallel. However, after doing some benchmarking, it was found much faster - with an admittedly more complex approach - with Pillow, c.f. https://github.com/Psycojoker/prosopopee/pull/137.

To be honest, I believe Python is not the correct language for this project. If I had the time, I'd probably rewrite it from scratch in another language. But I have found neither the time nor energy to do this in the last few years. Maybe in some years :) ?

**Comment by sebsto** [comment]
I understand all that perfectly - thank you for sharing the details. I was thinking about the same, maybe a rewrite in another programming language and leverage the heavy work that has been done on the templates.  If I have time ....

**Comment by QSchulz** [comment]
Could you please give a try to the code in the #28 pull request?

```
git clone https://github.com/QSchulz/recitale -b dev/cache-multiprocessing-spawn
RECITALEPATH=${PWD}/recitale
# From wherever your gallery is
PYTHONPATH=${RECITALEPATH} python3 -m recitale.recitale build
```

I haven't checked the commands, so lemme know if they don't work.

**Comment by sebsto** [comment]
Thank you Quentin for the changes.  The dreaded `EOFError` is now gone, but I receive a date related error that was not there.

```
➜  web_album git clone https://github.com/QSchulz/recitale -b dev/cache-multiprocessing-spawn
Cloning into 'recitale'...
remote: Enumerating objects: 5578, done.
remote: Counting objects: 100% (740/740), done.
remote: Compressing objects: 100% (212/212), done.
remote: Total 5578 (delta 578), reused 641 (delta 515), pack-reused 4838
Receiving objects: 100% (5578/5578), 61.41 MiB | 2.20 MiB/s, done.
Resolving deltas: 100% (3496/3496), done.

➜  web_album RECITALEPATH=${PWD}/recitale

➜  web_album echo $RECITALEPATH 
/Users/stormacq/Desktop/web_album/recitale

➜  web_album cd working 

➜  working ls
build         colombia2018  settings.yaml usa2017

➜  working rm -rf build 

➜  working rm -rf .recitale_cache 

➜  working PYTHONPATH=${RECITALEPATH} python3 -m recitale.recitale build
2023-08-30 07:14:53,813 WARNING (recitale.py:118) - Video: I can't locate the ffmpeg binary, please install the 'ffmpeg' package.
2023-08-30 07:14:53,813 WARNING (recitale.py:123) - Video: I won't be able to encode video and I will stop if I encounter a video to convert
2023-08-30 07:14:53,813 WARNING (recitale.py:132) - warning: If you want the rss, OpenGraph and/or the social network share to work, you need to specify the website url in root settings
Building galleries:  50%|██████████████████████████████████████████████▌                                              | 1/2 | ETA: 00:00
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/recitale.py", line 1032, in <module>
    main()
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/recitale.py", line 919, in main
    process_directory(
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/recitale.py", line 239, in process_directory
    build_gallery(settings, gallery_settings, gallery_path, parent_templates)
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/recitale.py", line 360, in build_gallery
    __build_gallery(settings, gallery_settings, gallery_path, gallery_path, template)
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/recitale.py", line 336, in __build_gallery
    html = template_to_render.render(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/homebrew/lib/python3.11/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/themes/exposure/templates/gallery-index.html", line 30, in top-level template code
    {% include "sections/" + section.type + ".html" %}
^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/themes/exposure/templates/sections/full-picture.html", line 39, in top-level template code
    <div class="datetime">{{ section.text.date|local_date("dd MMMM yyyy") }}</div>{% endif %}
^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/stormacq/Desktop/web_album/recitale/recitale/recitale.py", line 151, in local_date
    return format_date(date=value, format=date_format, locale=date_locale)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/babel/dates.py", line 681, in format_date
    locale = Locale.parse(locale)
             ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/babel/core.py", line 317, in parse
    raise TypeError(f"Unexpected value for identifier: {identifier!r}")
TypeError: Unexpected value for identifier: None
```

Looks like the `date_locale` parameter is None. Which is weird given https://github.com/QSchulz/recitale/blob/dev/cache-multiprocessing-spawn/recitale/recitale.py#L148

Looks like babel can not get the default local because environment variables are not set by default.

```
➜  working python3
Python 3.11.4 (main, Jul 25 2023, 17:36:13) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from babel.core import default_locale
>>> print(default_locale("LC_TIME"))
None
```

So I propose to add this to `recitale.py` 

```
*** recitale/recitale.orig.py	Wed Aug 30 07:34:23 2023
--- recitale/recitale.py	Wed Aug 30 07:37:46 2023
***************
*** 147,152 ****
--- 147,157 ----
      if date_locale is None:
          date_locale = default_locale("LC_TIME")
  
+     # double check because on some platforms, the environment variables are not positioned by default
+     if date_locale is None:
+         os.environ['LC_TIME'] = 'POSIX'
+         date_locale = default_locale('LC_TIME')
+ 
      def local_date(value, date_format="dd MMMM yyyy"):
          return format_date(date=value, format=date_format, locale=date_locale)
```

With this change, I managed to generate an album !  

Thank you

**Comment by QSchulz** [comment]
I created #29 for this locale issue. Can you please review/test again?

As I like to give proper attribution, I added your name + mail address in the commit log. Please let me know if you'd rather not be mentioned (or with which alternate identity).

**Comment by sebsto** [comment]
Thank you for the attribution - happy to keep it like you did.

**Comment by sebsto** [comment]
Nice parallel operations on a 10 cores machine 
<img width="579" alt="image" src="https://github.com/recitale/recitale/assets/401798/9952d572-c30e-40ac-9ddb-aa0ed38b1924">
